### PR TITLE
[Cocoa] Control whether PDF.js is included via a build-time setting

### PR DESCRIPTION
**Pull request by emw-apple**
#### d4387ce981a93c7bad97ff26700e5428c952e64a
<pre>
[Cocoa] Control whether PDF.js is included via a build-time setting
<a href="https://bugs.webkit.org/show_bug.cgi?id=241117">https://bugs.webkit.org/show_bug.cgi?id=241117</a>

Reviewed by Brent Fulgham.

Revert &quot;REGRESSION(r289850): Bundle size has increased&quot;
This reverts commit 3d8a962d9a56367e826fa968b4f1b4c8851b8982.

Expose a build-webkit feature flag `--[no-]pdfjs` which controls an
ENABLE_PDFJS build setting. This aligns with the GTK and WPE build
settings added in <a href="https://bugs.webkit.org/show_bug.cgi?id=238023.">https://bugs.webkit.org/show_bug.cgi?id=238023.</a>

On Xcode-based platforms, setting ENABLE_PDFJS causes pdfjs* resources
to be copied into WebCore, and ENABLE_PDFJS=1 to be defined.

* Source/WTF/wtf/PlatformEnable.h: For Cocoa platforms, disable
  ENABLE_PDFJS by default. It is conditionally enabled in xcconfigs.
* Source/WebCore/Configurations/WebCore.xcconfig: Recognize ENABLE_PDFJS
  setting. Define ENABLE_PDFJS as a preprocessor macro.
* Source/WebCore/Modules/pdfjs-extras/adwaita/style.css: Renamed from Source/WebCore/Modules/pdfjs/extras/adwaita/style.css.
* Source/WebCore/Modules/pdfjs-extras/cocoa/arrow.down.circle@2x.png: Renamed from Source/WebCore/Modules/pdfjs/extras/cocoa/arrow.down.circle@2x.png.
* Source/WebCore/Modules/pdfjs-extras/cocoa/minus.magnifyingglass@2x.png: Renamed from Source/WebCore/Modules/pdfjs/extras/cocoa/minus.magnifyingglass@2x.png.
* Source/WebCore/Modules/pdfjs-extras/cocoa/plus.magnifyingglass@2x.png: Renamed from Source/WebCore/Modules/pdfjs/extras/cocoa/plus.magnifyingglass@2x.png.
* Source/WebCore/Modules/pdfjs-extras/cocoa/style.css: Renamed from Source/WebCore/Modules/pdfjs/extras/cocoa/style.css.
* Source/WebCore/Modules/pdfjs-extras/content-script.js: Renamed from Source/WebCore/Modules/pdfjs/extras/content-script.js.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: Add copy file phases
  for PDF.js source and extras. Rather than copying the entire `pdfjs`
  directory, copy members individually to avoid copying README.webkit or
  PdfJSFiles.cmake into WebCore&apos;s bundle.
* Source/WebKit/Configurations/BaseTarget.xcconfig: Define ENABLE_PDFJS
  as a preprocessor macro.
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig: Define
  ENABLE_PDFJS as a preprocessor macro.
* Tools/Scripts/webkitperl/FeatureList.pm: Add pdfjs feature.

Canonical link: <a href="https://commits.webkit.org/252343@main">https://commits.webkit.org/252343@main</a>
</pre>

### CODE REVIEW

**Comment by emw-apple** [comment]
Combined with https://github.com/WebKit/WebKit/pull/2004, you can build with PDF.js enabled using

    make BUILD_WEBKIT_OPTIONS=--pdfjs

in addition to `build-webkit --pdfjs`.

If you're building in Xcode directly, add this line to `LocalOverrides.xcconfig`:

    ENABLE_PDFJS=ENABLE_PDFJS

File: Source/WTF/wtf/PlatformEnable.h
Comment (emw-apple) [inline Source/WTF/wtf/PlatformEnable.h]: One downside to moving this preprocessor macro logic to xcconfigs is that we don't have a single "Features" xcconfig that is used by every target. So, when building with `--pdfjs`, `ENABLE(PDFJS)` will only be true in WebCore, WebKit, and WebKitLegacy, because those projects are configured to recognize the build setting. It'll be false everywhere else.

File: Source/WTF/wtf/PlatformEnable.h
Comment (nt1m) [inline Source/WTF/wtf/PlatformEnable.h]: I just filed https://bugs.webkit.org/show_bug.cgi?id=242263 for this, can we reference this instead?

**Review by nt1m — COMMENTED** [review]
Thanks for looking into this. I don't know much about the build system, so I'll let others review.

I'm noticing individual files are being specified, is it possible to specify the whole PDF.js third party folder instead to ease-up PDF.js updates? Not a deal breaker, but would be nice to have.

**Comment by emw-apple** [comment]
@nt1m: 
> I'm noticing individual files are being specified, is it possible to specify the whole PDF.js third party folder instead to ease-up PDF.js updates? Not a deal breaker, but would be nice to have.

From the middle commit:

> Rather than copying the entire `pdfjs` directory, copy members
individually to avoid copying README.webkit or PdfJSFiles.cmake into
WebCore's bundle.

Let me know if that doesn't seem like the correct trade off.

BTW, I intend to squash these commits before landing, in case that's not clear :)

**Comment by nt1m** [comment]
> From the middle commit:

Ah sorry I missed the message, thanks for clarifying.

> Rather than copying the entire pdfjs directory, copy members
    individually to avoid copying README.webkit or PdfJSFiles.cmake into
    WebCore's bundle.

> Let me know if that doesn't seem like the correct trade off.

Thinking more, this might be a good idea if we want to exclude files for bundle size reasons. @brentfulgham WDYT about this particular trade-off? Do we value easier updates (that can be done by anyone including GTK folks) or picking what we want from the bundle?

Ideally we'd have the best of both worlds, e.g. a blocklist rather than allowlist, but not sure that's possible in XCode world.

**Comment by emw-apple** [comment]
> Do we value easier updates (that can be done by anyone including GTK folks) or picking what we want from the bundle?

As it's currently written, the entire `pdfjs/build` and `pdfjs/web` directories are copied (they are "directory references" in Xcode parlance). So additions/removals _within_ those directories do not require pbxproj changes. We'll only have to modify the project for things that are added or removed from top-level pdfjs directory.

**Review by brentfulgham — APPROVED** [review]
Much cleaner!

File: Source/WTF/wtf/PlatformEnable.h
Comment (brentfulgham) [inline Source/WTF/wtf/PlatformEnable.h:893]: Nit: Could you file a radar (or Bugzilla) for this task so we can be sure to re-enable it here?

Comment (nt1m) [reply Source/WTF/wtf/PlatformEnable.h:893]: Bugzilla: https://bugs.webkit.org/show_bug.cgi?id=242263
Radar: rdar://problem/96693948

**Pull request merged**
